### PR TITLE
Backport AutoValue integration

### PR DESCRIPTION
**Pull request by swegner**

### CODE REVIEW

**Comment by swegner** [comment]
R: @lukecwik 
CC: @tgroh

**Comment by swegner** [comment]
Travis failure seems to be a flake:

> Downloaded: https://repo.maven.apache.org/maven2/com/google/collections/google-collections/1.0/google-collections-1.0.jar (625 KB at 960.9 KB/sec)

> [INFO] Changes detected - recompiling the module!

> [INFO] Compiling 472 source files to /Users/travis/build/GoogleCloudPlatform/DataflowJavaSDK/sdk/target/classes



> No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.

> The build has been terminated

PTAL @lukecwik

**Comment by davorbonaci** [comment]
Do we need this backport?

**Comment by swegner** [comment]
re: @davorbonaci 
> Do we need this backport?

There have been a number of backported changes that needed manual merge work to un-AutoValue them, and @tgroh mentioned he has more in the pipeline. This isn't strictly required but should be relatively low-risk and make other backports easier.

**Comment by davorbonaci** [comment]
Sure; ok to proceed with caution.

**Comment by swegner** [comment]
I've addressed all feedback so far. Please take another look. @lukecwik

File: pom.xml [hunk 1]
```diff
@@ -205,6 +205,11 @@
               </goals>
             </execution>
           </executions>
+        <configuration>
```
Comment (lukecwik) [inline pom.xml]: Please fix indentation

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/worker/IsmFormat.java [hunk 2]
```diff
@@ -838,59 +743,29 @@ protected long getEncodedElementByteSize(KeyPrefix value, Coder.Context context)
    *   <li>0x01 (version key as a single byte)</li>
    * </ul>
    */
-  static class Footer {
+  @AutoValue
+  abstract static class Footer {
```
Comment (lukecwik) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/worker/IsmFormat.java]: public?

Comment (swegner) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/worker/IsmFormat.java]: No changes in visibility as part of the PR. Exposing as public was necessary in BEAM due to package restructuring. See: https://github.com/apache/incubator-beam/commit/e0b11311b287bdb8a303854676bf0f3e37d559c8

File: pom.xml (hunk 1 above)
Comment (swegner) [inline pom.xml]: Note: this has already been addressed in BEAM via https://github.com/apache/incubator-beam/commit/119812a66bceb9e8b91dcd7afb26bc9b7ee5eee7

**Comment by swegner** [comment]
I've addressed all feedback so far. Please take another look. @lukecwik

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/worker/IsmFormat.java (hunk 2 above)
Comment (lukecwik) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/worker/IsmFormat.java]: You added 4 spaces instead of 2

**Comment by swegner** [comment]
I've addressed all feedback so far. Please take another look. @lukecwik

**Comment by lukecwik** [comment]
LGTM

**Comment by lukecwik** [comment]
Waiting on travis and then will merge.

**Comment by lukecwik** [comment]
Can you sync your PR since there is a merge conflict now?

**Comment by swegner** [comment]
done. @lukecwik

**Pull request merged**